tests/Codec/LazyBSONCodecLibraryTest.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace MongoDB\Tests\Codec;
+
+use Generator;
+use MongoDB\BSON\Document;
+use MongoDB\BSON\PackedArray;
+use MongoDB\Codec\LazyBSONCodecLibrary;
+use MongoDB\Model\LazyBSONArray;
+use MongoDB\Model\LazyBSONDocument;
+use MongoDB\Tests\TestCase;
+
+class LazyBSONCodecLibraryTest extends TestCase
+{
+    public static function provideDecodedData(): Generator
+    {
+        $array = [
+            'bar',
+            ['foo' => 'bar'],
+            [0, 1, 2],
+        ];
+        $document = (object) [
+            'string' => 'bar',
+            'document' => (object) ['foo' => 'bar'],
+            'array' => [0, 1, 2],
+        ];
+
+        yield 'LazyBSONArray' => [
+            'expected' => PackedArray::fromPHP($array),
+            'value' => new LazyBSONArray($array),
+        ];
+
+        yield 'LazyBSONDocument' => [
+            'expected' => Document::fromPHP($document),
+            'value' => new LazyBSONDocument($document),
+        ];
+
+        yield 'array' => [
+            'expected' => [PackedArray::fromPHP($array)],
+            'value' => [new LazyBSONArray($array)],
+        ];
+
+        yield 'hash' => [
+            'expected' => ['foo' => PackedArray::fromPHP($array)],
+            'value' => ['foo' => new LazyBSONArray($array)],
+        ];
+
+        yield 'object' => [
+            'expected' => (object) ['foo' => PackedArray::fromPHP($array)],
+            'value' => (object) ['foo' => new LazyBSONArray($array)],
+        ];
+    }
+
+    public static function provideEncodedData(): Generator
+    {
+        $packedArray = PackedArray::fromPHP([
+            'bar',
+            ['foo' => 'bar'],
+            [0, 1, 2],
+        ]);
+        $document = Document::fromPHP([
+            'string' => 'bar',
+            'document' => ['foo' => 'bar'],
```

Comment:
I noticed the object casts are omitted here, but they'd also serve no purpose since you're converting this directly to BSON.